ra/SAPHana

```diff
@@ -2804,38 +2879,45 @@ function saphana_promote_clone() {
             # promote on the replica side...
             #
             hana_sync=$(get_SRHOOK "$sr_name" "$NODENAME")
-            case "$hana_sync" in
-                SOK )
-                    super_ocf_log info "ACT: !!!!!!! Promote REPLICA $SID-$InstanceName to be primary. !!!!!!"
-                    LPTloc=$(date '+%s')
-                    # lpa_set_lpt 20 $remoteNode
-                    lpa_set_lpt $LPTloc "$NODENAME"
-                    lpa_push_lpt $LPTloc
-                    # TODO: Get rid of the su by using a new interface:
-                    # SAPSYSTEMNAME=SLE /usr/sap/SLE/HDB00/HDBSettings.sh hdbnsutil -sr_takeover ...
-                    # TODO: Check beginning from which SPS does SAP support HDBSettings.sh?
-                    # TODO: Limit the runtime of hdbnsutil -sr_takeover ????
-                    # SAP_CALL
-                    set_hana_attribute "$NODENAME" "T" "${ATTR_NAME_HANA_SRACTION[@]}"
-                    set_hana_attribute "$NODENAME" "T" "${ATTR_NAME_HANA_SRACTION_HISTORY[@]}"
-                    HANA_CALL --timeout inf --use-su --cmd "hdbnsutil -sr_takeover"
-                    set_hana_attribute "$NODENAME" "-" "${ATTR_NAME_HANA_SRACTION[@]}"
-                    #
-                    # now gain check, if we are primary NOW
-                    #
-                    # TODO: PRIO3: check, if we need to destinguish between HANA_STATE_PRIMARY, HANA_STATE_SECONDARY, HANA_STATE_DEFECT
-                    #
-                    if check_for_primary; then
-                        rc=$OCF_SUCCESS;
-                    else
-                        rc=$OCF_FAILED_MASTER
-                    fi
-                    ;;
-                * )
-                    super_ocf_log err "ACT: HANA SYNC STATUS IS NOT 'SOK' SO THIS HANA SITE COULD NOT BE PROMOTED"
-                    rc=$OCF_ERR_GENERIC
-                    ;;
-            esac
+
+            if ocf_is_true ${IndexserverFailureTakeover}; then
+                indexserver_state_remote=$(get_hana_attribute "${remoteNode}" "${ATTR_NAME_HANA_INDEXSERVER[@]}")
+            else
+                indexserver_state_remote=''
+            fi
+
+            # Alow a takeover to be initiated when either the HSR sync state is SOK,
+            # or additionally when the primary indexserver status is available and unhealthy.
+            if [[ ${hana_sync} == "SOK" ]] || [[ ${indexserver_state_remote} =~ starting|stopping|down ]]; then
```

Comment:
It is NOT allowed to takeover, if SOK is set. Also in indexserver outage situations this is extremely dangerous and will lead most likely into data in-constancy. Also this solution does not address multiple index servers (e.g. multi-tenant).

Reply:
I'm confused here, why is a takeover not allowed when SOK is set? In the function `saphana_promote_clone` this is the vital requirement for a regular takeover.
The original lines 2807/2808 are a `case` conditional that runs the content when SOK is set. All I changed is the `case` to an `if/else` and added the primary indexserver state info as an extra (optional) conditional to the existing sync state = SOK. 
The reason is that at the moment when the promote function is executed on the secondary node, the sync state will already have changed to SFAIL in case of a failed primary indexserver, but should still proceed with a takeover. At this point the SAPHana resource on the primary node will have been failed by the cluster as part of `saphana_monitor_primary`, and the secondary is running the promotion.

Regarding multi-tenancy I may have overlooked the related difference in the existing `srHook/susChkSrv.py` that deals with this. Would you please be so kind and point me to the detail for reference?

Reply:
@ja9fuchs It is exactly the opposite! Your code does overrule the code checking for SOK by on OR (||). 
So you would also go to the positve if-branch, if hana_sync != SOK BUT indexserver_state_remote} =~ starting|stopp|... But that is dangerous and unacceptable!

Reply:
I see your point, this is an issue if the system replication was already out of sync before and the takeover request is triggered by a situation unrelated to just an indexserver failure. I'll have to think about that again.